### PR TITLE
Move supporting matrix generator assets out of `eng/common`

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
This PR contributes to addressing:
- #4995 

by moving out of `eng/common/scripts/job-matrix` assets that should not be mirrored to language repos. These assets include documentation, samples, and tests.

## Changes made

- Moved contents of `eng/common/scripts/job-matrix/README.md` to newly introduced `doc/common/matrix_generator.md`;
- Replaced contents of `eng/common/scripts/job-matrix/README.md` with a pointer to `doc/common/matrix_generator.md`;
- Moved contents of `eng/common/scripts/job-matrix/scripts` and `eng/common/scripts/job-matrix/tests` to corresponding directories in newly introduced directory of 
`eng/common-tests/matrix-generator`.

Note the introduced directory and doc file use `matrix generator` as name, not `job matrix`. This matches the terminology used in the documentation.

## Changes NOT made

- The contents of files moved from `scripts` and `tests` directories have not been updated at all. The files have been moved unchanged.
- The contents of the moved `README.md` have not been updated, even though they should, as they refer to paths that have changed as a result of the move. To keep the PR diff clean, I will instead update them in a separate PR:
  - #5023

### CODE REVIEW

File: doc/common/matrix_generator.md [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+```
+
+The matrix is stored as a one-dimensional array, with a row-major indexing scheme (e.g. `(2, 1, 0, 1, 0)`).
+
+## Testing
+
+The matrix functions can be tested using [pester](https://pester.dev/). The test command must be run from within the tests directory.
+
+```
+$ cd tests
```
Comment (benbp) [inline doc/common/matrix_generator.md:551]:
```suggestion
$ cd <azure sdk tools repo root>/eng/common-tests/matrix-generator
```

File: doc/common/matrix_generator.md [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+$ Invoke-Pester
+
+Starting discovery in 3 files.
+Discovery finished in 75ms.
+[+] /home/ben/sdk/azure-sdk-tools/eng/common/scripts/job-matrix/tests/job-matrix-functions.filter.tests.ps1 750ms (309ms|428ms)
+[+] /home/ben/sdk/azure-sdk-tools/eng/common/scripts/job-matrix/tests/job-matrix-functions.modification.tests.ps1 867ms (250ms|608ms)
+[+] /home/ben/sdk/azure-sdk-tools/eng/common/scripts/job-matrix/tests/job-matrix-functions.tests.ps1 2.71s (725ms|1.93s)
```
Comment (benbp) [inline doc/common/matrix_generator.md:558]:
```suggestion
[+] /home/ben/sdk/azure-sdk-tools/eng/common-tests/matrix-generator/tests/job-matrix-functions.filter.tests.ps1 750ms (309ms|428ms)
[+] /home/ben/sdk/azure-sdk-tools/eng/common-tests/matrix-generator/tests/job-matrix-functions.modification.tests.ps1 867ms (250ms|608ms)
[+] /home/ben/sdk/azure-sdk-tools/eng/common-tests/matrix-generator/tests/job-matrix-functions.tests.ps1 2.71s (725ms|1.93s)
```

**Review by benbp — APPROVED** [review]
Minor updates to the doc to point to the new test location. Otherwise LGTM, thanks!

File: doc/common/matrix_generator.md [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+
+For a single matrix, you can include the `/eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml` template in a pipeline (see /eng/common/scripts/job-matrix/samples/matrix-test.yml for a full working example):
+
+```
+jobs:
+  - template: /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
+    parameters:
+      MatrixConfigs:
+        - Name: base_product_matrix
+          Path: eng/common/scripts/job-matrix/samples/matrix.json
```
Comment (weshaggard) [inline doc/common/matrix_generator.md:42]: Looks like there are some paths you also need to update.

Comment (konrad-jamrozik) [reply doc/common/matrix_generator.md:42]: @weshaggard thanks for pointing this out! I will review and fix the content of the `matrix_generator.md` in a pending, stacked follow-up PR:
- #5023

Comment (konrad-jamrozik) [reply doc/common/matrix_generator.md:42]: TIL: we have link verification :D

**Pull request merged**